StRoot/StFcsWaveformFitMaker/PeakAna.cxx

```diff
@@ -0,0 +1,1336 @@
+{
+	mSearch.SetWindow(peak,width);
+  ResetPeak();//reset since old found peak should be invalid
+}
+
+void PeakAna::SetPeak(const Int_t peakpoint, const Double_t peakx )
+{ mFoundPeak.mP_Peak=peakpoint; mFoundPeak.mPeakX=peakx; }
+void PeakAna::SetWindow(const Int_t start, const Int_t end )
+{ mFoundPeak.SetWindow(start,end); }
+
+void PeakAna::SetWindow(PeakWindow window)
+{ mFoundPeak=window; }
+
+void PeakAna::SetTunnelScale(Double_t value)
+{
+  if( value>=0 ){mTunnelScale = value;}
+}
+void PeakAna::SetTunnelSigma(Double_t value)
+{
+  if( value>0 ){mTunnelSigma = value;}
+}
+void PeakAna::SetTunnelPars(Double_t scale, Double_t sigma)
+{
+  SetTunnelScale(scale);
+  SetTunnelSigma(sigma);
+}
+void PeakAna::SetTunnelThreshold(Double_t value)
+{
+  if( value<=1 ){ mTunnelThreshold=value; }
+}
+
+
+TGraph* PeakAna::ConvertHistToGraph(TH1* hist, UInt_t numavgs)
+{
+  Int_t nbins = hist->GetNbinsX();
+  //Only consider averaging when number of bins to average are greater than or equal to 1 or less than the number of bins
+  if( numavgs<1 || static_cast<Int_t>(numavgs)>nbins ){return 0;}//strictly greater than since GetNbinsX()+1 is overflow
+  TGraph* AvgGr = new TGraph();
+  for( Int_t ibin=1; ibin<=nbins; ibin+=numavgs ){
+    Double_t sum=0;
+    UInt_t counter=0;//In case number of averages is not same as how many sums were performed
+    for(Int_t i=ibin; i<ibin+static_cast<Int_t>(numavgs) && i<=nbins; ++i ){ sum += hist->GetBinContent(i); ++counter; }
+    Double_t avg = sum/static_cast<Double_t>(counter);
+    //std::cout << std::endl<< "|sum:"<<sum << "|avg:"<<avg <<"|val:"<<hist->GetBinContent(ibin) << std::endl;
+	  //Set graph's point to center of x range
+    Double_t xlow = hist->GetBinLowEdge(ibin);
+    Double_t xhigh = hist->GetBinLowEdge(ibin+counter);
+    //AvgGr->SetPoint(AvgGr->GetN(),static_cast<Double_t>(ibin)+static_cast<Double_t>(numavgs)*0.5,avg);//This is for knowing which bin
+    AvgGr->SetPoint(AvgGr->GetN(),(xlow+xhigh)/2.0,avg);//This is for knowing the x-value
+  }
+  return AvgGr;
+}
+
+TGraph* PeakAna::ConvertPeaksToGraph(const std::vector<PeakWindow> &Peaks)
+{
+  if( Peaks.size()==0 ){return 0;}
+  TGraph* graph = new TGraph();
+  for( UInt_t ipeak=0; ipeak<Peaks.size(); ++ipeak )
+  {
+    graph->SetPoint( ipeak,Peaks.at(ipeak).mPeakX,Peaks.at(ipeak).MidPoint() );
+  }
+  return graph;
+}
+
+void PeakAna::ConvertPeaksToGraph()
+{
+  TGraph* graph = new TGraph();
+  for( UInt_t ipeak=0; ipeak<mPeaks.size(); ++ipeak )
+  {
+    graph->SetPoint( ipeak,mPeaks.at(ipeak).mPeakX,mPeaks.at(ipeak).MidPoint() );
+  }
+  this->SetData(graph);
+  this->ForceInternal();//new graph replaces old one so needs to be deleted
+  return;
+}
+
+PeakAna* PeakAna::ConvertPeaksToAna(const PeakAna &Ana)
+{
+  TGraph* graph = new TGraph();
+  for( Int_t ipeak=0; ipeak<Ana.NPeaks(); ++ipeak )
+  {
+    graph->SetPoint(ipeak, Ana.GetPeak(ipeak).mPeakX, Ana.GetPeak(ipeak).MidPoint() );
+  }
+  PeakAna* NewAna = new PeakAna(Ana,graph);
```

Comment:
This can cause a memory leak. Why not return the object?

Reply:
I am not sure whether you are referring to *graph* or *NewAna*. I don't think *graph* will give a memory leak since the next line calls `NewAna->ForceInternal()`. This means that the newly created graph object is "owned" by NewAna and it will delete that newly created graph when it is destroyed or *SetData()* is called. There is no way to change mInternalSignal to false after that so NewAna will never think that it doesn't own *graph*. If you are referring to NewAna then I agree it will if someone forgets to delete it. I don't want to return the graph object itself because if someone does `PeakAna* newana = new PeakAna(OldAna,ConvertPeaksToAna(OldAna))` then the way this constructor works it will think the graph object is external to it and it won't delete it and that will be a memory leak. Which means that the caller will have to remember to always call `newana->ForceInternal()` in this situation. Here I called it automatically to prevent such a situation so that's why I don't return the graph object.

Reply:
Actually, I meant you could return the NewAna object by value rather than by pointer, not the graph object. I am a bit lost on whether PeakAna owns the graph object or not and under what conditions... Hopefully, this complex (for me) logic is correct and there are no memory leaks.

Reply:
I thought about returning by value but my understanding is that when you return by value it returns a copy of the object not the object created in the function itself. This is not something I want for a large object like PeakAna. Also, if you try to get around this by returning a reference instead then the returned reference is to the object created in the function which has now been invalidated and hence does not exist. This is why I return the pointer to the new object. Is my understanding incorrect and is there a safe way to do this without copying the object?

Reply:
done